### PR TITLE
PR#2107 improved - Concox CRX1 contrib

### PR DESCRIPTION
**Pull request by guedou**
This PR improves #2107 and includes changes requested by @gpotter2.

### CODE REVIEW

File: scapy/contrib/concox.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    0x04: "Strong Signal",
+}
+
+LANGUAGE = {
+    0x01: "Chinese",
+    0x02: "English",
+}
+
+
+class BCDStrFixedLenField(StrFixedLenField):
```
[the remark this answers is not in the document]
Comment (guedou) [reply scapy/contrib/concox.py]: Which one? I tried and did not find the correct alternative.

File: scapy/contrib/concox.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    name = "CRX1 New Packet Content"
+    fields_desc = [
+        XByteEnumField('protocol_number', 0x12, PROTOCOL_NUMBERS),
+        # Login
+        ConditionalField(
+            BCDStrFixedLenField('terminal_id', '00000000', length=8), lambda
+            pkt: len(pkt.original) > 5 and pkt.protocol_number == 0x01),
+        # GPS Location
+        ConditionalField(
+            ByteField('year', 0x00), lambda pkt: len(pkt.original) > 5 and pkt.
```
[the remark this answers is not in the document]
Comment (guedou) [reply scapy/contrib/concox.py]: flake8 is OK with that. I don't want to spend time fixing this module.

File: scapy/contrib/concox.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+            protocol_number in (0x12, 0x16)),
+        ConditionalField(
+            ByteField('month', 0x01), lambda pkt: len(pkt.original) > 5 and pkt
+            .protocol_number in (0x12, 0x16)),
+        ConditionalField(
+            ByteField('day', 0x01), lambda pkt: len(pkt.original) > 5 and pkt.
+            protocol_number in (0x12, 0x16)),
+        ConditionalField(
+            ByteField('hour', 0x00), lambda pkt: len(pkt.original) > 5 and pkt.
```
[the remark this answers is not in the document]
Comment (guedou) [reply scapy/contrib/concox.py]: While not perfect, the current implementation seems to work. I think that's enough for a contrib module.

File: scapy/contrib/concox.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    def i2h(self, pkt, x):
+        if isinstance(x, bytes):
+            return binascii.b2a_hex(x)
+        return binascii.a2b_hex(x)
+
+
+class CoordinateField(IntField):
+    def i2h(self, pkt, x):
+        value = super(CoordinateField, self).i2h(pkt, x)
+        return round(value / 1800000, 6)
```
[the remark this answers is not in the document]
Comment (guedou) [reply scapy/contrib/concox.py]: Could you provide an example?

File: scapy/contrib/concox.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/contrib/concox.py]: `XStrFixedLenField` ?

File: scapy/contrib/concox.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/contrib/concox.py]:
```python
>>> f = ScalingField("latitude", 0, scaling=1/1800000, ndigits=6, fmt="!I")
>>> f.addfield(None, b"", 48.293024)
b'\x05.hS'
>>> f.getfield(None, b'\x05.hS')
(b'', 48.293024)
```

File: scapy/contrib/concox.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (guedou) [reply scapy/contrib/concox.py]: It does not work the same way.

File: scapy/contrib/concox.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (guedou) [reply scapy/contrib/concox.py]: Fixed.

**Comment by guedou** [comment]
Ready to be merged!

**Comment by p-l-** [comment]
Fixes #2107.

**Comment by guedou** [comment]
I squashed this myself as the PR was approved by @p-l-

**Pull request merged**